exotations/task_maps/task_map/src/JointJerkBackFiniteDiff.cpp

```diff
@@ -0,0 +1,101 @@
+ *    endorse or promote products derived from this software without specific
+ *    prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include "task_map/JointJerkBackFiniteDiff.h"
+
+REGISTER_TASKMAP_TYPE("JointJerkBackFiniteDiff", exotica::JointJerkBackFiniteDiff);
+
+namespace exotica
+{
+// Assign scene and initialize
+void JointJerkBackFiniteDiff::assignScene(Scene_ptr scene)
+{
+    scene_ = scene;
+    Initialize();
+}
+
+// Instantiate
+void JointJerkBackFiniteDiff::Instantiate(JointJerkBackFiniteDiffInitializer& init)
+{
+    init_ = init;
+}
+
+// Initialize Q_
+//   Q_ is a log of the previous three joint states.
+//   Q_.col(0) is the most recent joint state.
+//   initPrevJointState sets each column of Q_ to joint_state.
+void JointJerkBackFiniteDiff::initPrevJointState(Eigen::VectorXdRefConst joint_state)
+{
+    if (joint_state.rows() != N) throw_named("Wrong size for joint_state!");
+    Q_.resize(N, 3);
+    for (int i = 0; i < 3; i++)
+        Q_.col(i) = joint_state;
+}
+
+// Sets the previous joint state.
+//   In users script (python/c++) this must be called preceeding every time solve.solve() is called.
+//   Pushes all columns back and sets joint_state as most recent previous joint state.
+void JointJerkBackFiniteDiff::setPrevJointState(Eigen::VectorXdRefConst joint_state)
+{
+    if (joint_state.rows() != N) throw_named("Wrong size for joint_state!");
+    Q_.col(2) = Q_.col(1);
+    Q_.col(1) = Q_.col(0);
+    Q_.col(0) = joint_state;
+    Qe_ = Q_ * e;
+}
+
+// Initialize task map.
+//  Sets N (number of dofs), I_ used to set the Jacobian in update(), and e.
+//  e is the constant vector (-3 3 -1) used to approximate the joint jerk using backwards finite differencing.
+void JointJerkBackFiniteDiff::Initialize()
+{
+    N = scene_->getSolver().getNumControlledJoints();
+    I_ = Eigen::MatrixXd::Identity(N, N);
+    e << -3, 3, -1;
+}
+
+void JointJerkBackFiniteDiff::update(Eigen::VectorXdRefConst x, Eigen::VectorXdRef phi)
+{
+    phi = x + Qe_;
+}
+
+void JointJerkBackFiniteDiff::update(Eigen::VectorXdRefConst x, Eigen::VectorXdRef phi, Eigen::MatrixXdRef J)
+{
+    phi = x + Qe_;
```

Comment:
Check dimensionality for Phi and J... and throw if they aren't correct. We had too many bugs due to wrong sizes to not do it. Please change also all three files regarding this.